contracts/profit/src/state/buy_back.rs

```diff
@@ -96,13 +98,15 @@ impl SwapTask for BuyBack {
     where
         Visitor: CoinVisitor<Result = IterNext>,
     {
-        TryFind::try_find(&mut self.coins.iter(), |coin: &&CoinDTO<PaymentGroup>| {
+        let mut coins_iter: SliceIter<'_, CoinDTO<PaymentGroup>> = self.coins.iter();
+
+        TryFind::try_find(&mut coins_iter, |coin: &&CoinDTO<PaymentGroup>| {
             visitor
                 .visit(coin)
                 .map(|result: IterNext| matches!(result, IterNext::Stop))
         })
-        .map(|maybe_coin: Option<&CoinDTO<PaymentGroup>>| {
-            if maybe_coin.is_some() {
+        .map(|_| {
+            if coins_iter.as_slice().is_empty() {
                 IterState::Complete
             } else {
                 IterState::Incomplete
```

Comment:
the previous implementation seems to work ... just switch the `then` with the `else` blocks

Reply:
I can't agree with that because it can "find" the last element, in which it would've resolved the whole sequence. Doing that directly contradicts with the debug assertions in the calling component.

Reply:
Took the liberty of reverting locally and replacing `Option::is_some` with `Option::is_none`. Other tests begin to fail, thus not a wanted behavior.